69-common-kingslayer/technical_specification.md

```diff
@@ -0,0 +1,51 @@
+# OpenTelemetry in file services (Common Kingslayer)
+**Epic Type:** Implementation Epic
+
+Epic planning and implementation follow the
+[Epic Planning and Marathon SOP](https://docs.ghga-dev.de/main/sops/sop001_epic_planning.html).
+
+## Scope
+### Outline:
+
+Based on some previous exploration work, this epic aims to implement 1) OpenTelemetry instrumentation in the file services and 2) necessary OpenTelemetry functionality inside of hexkit and/or ghga-service-commons.
+In addition, a viable setup of Jaeger and its components needs to be devised and implemented.
+
```

Comment:
Again, specify whether you want to support spans that involve multiple file services. If yes, it makes sense to also extend it to other services like the claims repo or even the api gateway or auth adapter.

The scope could be reduced to implement only one or two larger spans that contain both sync and async request flow (probably those that we have been struggling with in the last release), but with reusable code/providers in hexkit/service-commons.

Reply:
Should this epic also quickly check alternatives to Jaeger, like Zipkin or Grafana Tempo?

Reply:
Would be good to specify which parts of the infrastructure should be covered by the instrumentation, e.g. S3, Mongo, sync and async calls. We want to see whether e.g. certain DB queries are very long.

Would be also good to show as an example how to trace certain code parts, e.h. calling a function that encodes or decodes a block of data.

Reply:
I wouldn't include exploration of other backends for now.
Unsure if those are just plug and play replacements with some extra config or if there's more involved in getting those to run.
Current expertise from exploration is with Jaeger and if that's good enough for now, I would go with that. Either way it's not a critical part of the infrastructure and we can keep exploring the alternatives in the backlog for now.
To me that seems like a mini epic and if we should deem Jaeger unfit for our use case down the road, then we can reevaluate.

Reply:
Spans should cross service boundaries, yes.
I'll add details for other services. 
I would still implement that in the FSB first as there's existing code/config from exploration and then would move on to other services that would benefit from that afterwards.

Reply:
> Would be good to specify which parts of the infrastructure should be covered by the instrumentation, e.g. S3, Mongo, sync and async calls. We want to see whether e.g. certain DB queries are taking too long.
> 
> Would be also good to show as an example how to trace certain code parts, e.h. calling a function that encodes or decodes a block of data.

The first part I can extract from the exploration requirements.

What do you mean by the second part, i.e. showing how to open a span around that and an instance of the associated trace?

Reply:
Ok. Maybe just mention that exploring other backends is not part of this epic but may be done later if we are not satisfied with Jaeger.

For the second part, yes, essentially just showing how to open some spans around such functions. Maybe also checking if opentelemetry-instrumentation-asyncio could be useful in this regard.